### ISSUE TITLE
CME saving world

### ISSUE THREAD

**Issue by Tommeh1** [GTNewHorizons/GT-New-Horizons-Modpack#13075]
### Your Pack Version

2.3.0

### Your Server

Private Server

### Java Version

Java 8

### Type of Server

Vanilla Forge

### Your Expectation

/

### The Reality
No further details as to what caused it were logged, this may require additional logging to be added somewhere.

```java
[17:49:16] [File IO Thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: java.util.ConcurrentModificationException
[17:49:16] [File IO Thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: at java.util.HashMap$HashIterator.nextNode(HashMap.java:1445)
[17:49:16] [File IO Thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: at java.util.HashMap$KeyIterator.next(HashMap.java:1469)
[17:49:16] [File IO Thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:29)
[17:49:16] [File IO Thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:376)
[17:49:16] [File IO Thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:31)
[17:49:16] [File IO Thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:376)
[17:49:16] [File IO Thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:31)
[17:49:16] [File IO Thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:376)
[17:49:16] [File IO Thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:31)
[17:49:16] [File IO Thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: at net.minecraft.nbt.NBTTagList.func_74734_a(NBTTagList.java:32)
[17:49:16] [File IO Thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:376)
[17:49:16] [File IO Thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:31)
[17:49:16] [File IO Thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:376)
[17:49:16] [File IO Thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:31)
[17:49:16] [File IO Thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: at net.minecraft.nbt.CompressedStreamTools.func_150663_a(CompressedStreamTools.java:149)
[17:49:16] [File IO Thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: at net.minecraft.nbt.CompressedStreamTools.func_74800_a(CompressedStreamTools.java:139)
[17:49:16] [File IO Thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: at net.minecraft.world.chunk.storage.AnvilChunkLoader.func_75821_a(AnvilChunkLoader.java:265)
[17:49:16] [File IO Thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: at net.minecraft.world.chunk.storage.AnvilChunkLoader.func_75814_c(AnvilChunkLoader.java:251)
[17:49:16] [File IO Thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: at net.minecraft.world.storage.ThreadedFileIOBase.func_75736_b(SourceFile:30)
[17:49:16] [File IO Thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: at net.minecraft.world.storage.ThreadedFileIOBase.run(SourceFile:23)
```

### Your Proposal

/

### Final Checklist

- [X] I have searched this issue tracker and there is nothing similar already. Posting on a closed issue saying the bug still exists will prompt us to investigate and reopen it once we confirm your report.
- [ ] I can reproduce this problem consistently by follow the exact steps I described above, or this does not need reproducing, e.g. recipe loophole.
- [ ] I have asked other people and they confirm they also have this problem by follow the exact steps I described above, or this does not need reproducing, e.g. recipe loophole.

**Comment by Glease** [comment]
huh weird. who is it breaking world save again this time around.

**Comment by Tommeh1** [comment]
> huh weird. who is it breaking world save again this time around.

If you've got any additional debugging to attach i'll gladly try it, given that i don't want this to occur on a restart or when saving is attempted after a crash, as this would then often result in a slight rollback.

**Comment by Glease** [comment]
yeah I understand the severity of the bug. now for a explanation on this bug.

The world process is done in 2 phase. 
1. data preparation phase: minecraft ask entities and tiles to save their states to an NBT tag. then forge ask mod to save arbitrary data not associated with specific tile entity and entity. 
2. data serialization phase: data is sent to a writer thread to save on disk. 

Basically the bug is caused by *something* retaining a reference to the data they have written in phase 1, and while phase 2 is happening, they changed this tag. Java is very unhappy with this as it can lead to undefined/very confusing behavior, so it decided to crash outright. Now the obvious problem is finding out who is mutating, but this can be a troublesome step, as the exception is thrown on the reader side (which did nothing wrong) and the culprit get away with no marks on it.

I naively believed tectech was the only culprit, turns out it isn't alone. The easier part, i.e. looking through code of `save arbitrary data not associated with specific tile entity and entity`, is done while sorting the previous ticket, so it's some tile entity or entity's fault. I'll ask fellow developers and see if they have any recollections on this.

**Comment by Glease** [comment]
Try replace hodgepodge with http://jenkins.usrv.eu:8080/job/Hodgepodge/lastSuccessfulBuild/artifact/build/libs/hodgepodge-2.2.4-2-g31a060e.jar and enable `debug.chunkSaveCMEDebug` in hodgepodge config

**Comment by Tommeh1** [comment]
> Try replace hodgepodge with http://jenkins.usrv.eu:8080/job/Hodgepodge/lastSuccessfulBuild/artifact/build/libs/hodgepodge-2.2.4-2-g31a060e.jar and enable `debug.chunkSaveCMEDebug` in hodgepodge config

Enabling that debug will cause a crash:
```java
[15:21:35] [main/TRACE] [mixin/]: Added class metadata for java/lang/Thread to metadata cache
[15:21:35] [main/TRACE] [mixin/]: catching
java.lang.ClassNotFoundException: The specified class 'com.llamalad7.mixinextras.injector.wrapoperation.Operation' was not found
	at org.spongepowered.asm.service.mojang.MixinServiceLaunchWrapper.getClassBytes(MixinServiceLaunchWrapper.java:519) ~[gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.service.mojang.MixinServiceLaunchWrapper.getClassNode(MixinServiceLaunchWrapper.java:592) ~[gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.ClassInfo.forName(ClassInfo.java:2005) [gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.ClassInfo.forType(ClassInfo.java:2059) [gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.ClassInfo.forDescriptor(ClassInfo.java:2038) [gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.MixinPreProcessorStandard.transformMethod(MixinPreProcessorStandard.java:752) [gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.MixinPreProcessorStandard.transform(MixinPreProcessorStandard.java:739) [gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.MixinPreProcessorStandard.attach(MixinPreProcessorStandard.java:310) [gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.MixinPreProcessorStandard.createContextFor(MixinPreProcessorStandard.java:280) [gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.MixinInfo.createContextFor(MixinInfo.java:1288) [gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.apply(MixinApplicatorStandard.java:292) [gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.TargetClassContext.apply(TargetClassContext.java:383) [gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.TargetClassContext.applyMixins(TargetClassContext.java:365) [gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.MixinProcessor.applyMixins(MixinProcessor.java:363) [gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClass(MixinTransformer.java:234) [gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClassBytes(MixinTransformer.java:202) [gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.Proxy.transform(Proxy.java:72) [gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at net.minecraft.launchwrapper.LaunchClassLoader.runTransformers(LaunchClassLoader.java:279) [launchwrapper-1.12.jar:?]
	at net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:176) [launchwrapper-1.12.jar:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418) [?:1.8.0_282]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351) [?:1.8.0_282]
	at net.minecraft.block.Block.func_149671_p(Block.java:355) [aji.class:?]
	at net.minecraft.init.Bootstrap.func_151354_b(SourceFile:355) [kl.class:?]
	at net.minecraft.server.MinecraftServer.main(MinecraftServer.java:1260) [MinecraftServer.class:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_282]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_282]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_282]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_282]
	at net.minecraft.launchwrapper.Launch.launch(Launch.java:135) [launchwrapper-1.12.jar:?]
	at net.minecraft.launchwrapper.Launch.main(Launch.java:28) [launchwrapper-1.12.jar:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_282]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_282]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_282]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_282]
	at cpw.mods.fml.relauncher.ServerLaunchWrapper.run(ServerLaunchWrapper.java:43) [forge-1.7.10-10.13.4.1614-1.7.10-universal.jar:?]
	at cpw.mods.fml.relauncher.ServerLaunchWrapper.main(ServerLaunchWrapper.java:12) [forge-1.7.10-10.13.4.1614-1.7.10-universal.jar:?]
[15:21:35] [main/WARN] [mixin/]: Error loading class: com/llamalad7/mixinextras/injector/wrapoperation/Operation (java.lang.ClassNotFoundException: The specified class 'com.llamalad7.mixinextras.injector.wrapoperation.Operation' was not found)
[15:21:35] [main/TRACE] [mixin/]: Added class metadata for com/llamalad7/mixinextras/injector/wrapoperation/Operation to metadata cache
[15:21:35] [main/ERROR] [LaunchWrapper/]: Unable to launch
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_282]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_282]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_282]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_282]
	at net.minecraft.launchwrapper.Launch.launch(Launch.java:135) [launchwrapper-1.12.jar:?]
	at net.minecraft.launchwrapper.Launch.main(Launch.java:28) [launchwrapper-1.12.jar:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_282]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_282]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_282]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_282]
	at cpw.mods.fml.relauncher.ServerLaunchWrapper.run(ServerLaunchWrapper.java:43) [forge-1.7.10-10.13.4.1614-1.7.10-universal.jar:?]
	at cpw.mods.fml.relauncher.ServerLaunchWrapper.main(ServerLaunchWrapper.java:12) [forge-1.7.10-10.13.4.1614-1.7.10-universal.jar:?]
Caused by: java.lang.NoClassDefFoundError: net/minecraft/nbt/NBTTagCompound
	at net.minecraft.block.Block.func_149671_p(Block.java:355) ~[aji.class:?]
	at net.minecraft.init.Bootstrap.func_151354_b(SourceFile:355) ~[kl.class:?]
	at net.minecraft.server.MinecraftServer.main(MinecraftServer.java:1260) ~[MinecraftServer.class:?]
	... 12 more
Caused by: java.lang.ClassNotFoundException: net.minecraft.nbt.NBTTagCompound
	at net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:191) ~[launchwrapper-1.12.jar:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418) ~[?:1.8.0_282]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351) ~[?:1.8.0_282]
	at net.minecraft.block.Block.func_149671_p(Block.java:355) ~[aji.class:?]
	at net.minecraft.init.Bootstrap.func_151354_b(SourceFile:355) ~[kl.class:?]
	at net.minecraft.server.MinecraftServer.main(MinecraftServer.java:1260) ~[MinecraftServer.class:?]
	... 12 more
Caused by: org.spongepowered.asm.mixin.transformer.throwables.MixinTransformerError: An unexpected critical error was encountered
	at org.spongepowered.asm.mixin.transformer.MixinProcessor.applyMixins(MixinProcessor.java:392) ~[gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClass(MixinTransformer.java:234) ~[gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClassBytes(MixinTransformer.java:202) ~[gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.Proxy.transform(Proxy.java:72) ~[gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at net.minecraft.launchwrapper.LaunchClassLoader.runTransformers(LaunchClassLoader.java:279) ~[launchwrapper-1.12.jar:?]
	at net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:176) ~[launchwrapper-1.12.jar:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418) ~[?:1.8.0_282]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351) ~[?:1.8.0_282]
	at net.minecraft.block.Block.func_149671_p(Block.java:355) ~[aji.class:?]
	at net.minecraft.init.Bootstrap.func_151354_b(SourceFile:355) ~[kl.class:?]
	at net.minecraft.server.MinecraftServer.main(MinecraftServer.java:1260) ~[MinecraftServer.class:?]
	... 12 more
Caused by: org.spongepowered.asm.mixin.throwables.ClassMetadataNotFoundException: com.llamalad7.mixinextras.injector.wrapoperation.Operation
	at org.spongepowered.asm.mixin.transformer.MixinPreProcessorStandard.transformMethod(MixinPreProcessorStandard.java:754) ~[gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.MixinPreProcessorStandard.transform(MixinPreProcessorStandard.java:739) ~[gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.MixinPreProcessorStandard.attach(MixinPreProcessorStandard.java:310) ~[gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.MixinPreProcessorStandard.createContextFor(MixinPreProcessorStandard.java:280) ~[gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.MixinInfo.createContextFor(MixinInfo.java:1288) ~[gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.apply(MixinApplicatorStandard.java:292) ~[gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.TargetClassContext.apply(TargetClassContext.java:383) ~[gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.TargetClassContext.applyMixins(TargetClassContext.java:365) ~[gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.MixinProcessor.applyMixins(MixinProcessor.java:363) ~[gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClass(MixinTransformer.java:234) ~[gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClassBytes(MixinTransformer.java:202) ~[gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at org.spongepowered.asm.mixin.transformer.Proxy.transform(Proxy.java:72) ~[gtnhmixins-2.1.13.jar:0.8.5-GTNH-2]
	at net.minecraft.launchwrapper.LaunchClassLoader.runTransformers(LaunchClassLoader.java:279) ~[launchwrapper-1.12.jar:?]
	at net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:176) ~[launchwrapper-1.12.jar:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418) ~[?:1.8.0_282]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351) ~[?:1.8.0_282]
	at net.minecraft.block.Block.func_149671_p(Block.java:355) ~[aji.class:?]
	at net.minecraft.init.Bootstrap.func_151354_b(SourceFile:355) ~[kl.class:?]
	at net.minecraft.server.MinecraftServer.main(MinecraftServer.java:1260) ~[MinecraftServer.class:?]
	... 12 more
[15:21:35] [main/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: java.lang.reflect.InvocationTargetException
[15:21:35] [main/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[15:21:35] [main/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[15:21:35] [main/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[15:21:35] [main/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at java.lang.reflect.Method.invoke(Method.java:498)
[15:21:35] [main/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at cpw.mods.fml.relauncher.ServerLaunchWrapper.run(ServerLaunchWrapper.java:43)
[15:21:35] [main/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at cpw.mods.fml.relauncher.ServerLaunchWrapper.main(ServerLaunchWrapper.java:12)
[15:21:35] [main/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: Caused by: cpw.mods.fml.relauncher.FMLSecurityManager$ExitTrappedException
[15:21:35] [main/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at cpw.mods.fml.relauncher.FMLSecurityManager.checkPermission(FMLSecurityManager.java:25)
[15:21:35] [main/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at java.lang.SecurityManager.checkExit(SecurityManager.java:761)
[15:21:35] [main/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at java.lang.Runtime.exit(Runtime.java:107)
[15:21:35] [main/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at java.lang.System.exit(System.java:973)
[15:21:35] [main/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.launchwrapper.Launch.launch(Launch.java:138)
[15:21:35] [main/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.launchwrapper.Launch.main(Launch.java:28)
[15:21:35] [main/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	... 6 more
```

**Comment by Glease** [comment]
Huh maybe that utility class wasn't in the old gtnhmixin. try upgrade gtnhmixins to unimixin found in 2.3.2 zip, or just use this build instead: http://jenkins.usrv.eu:8080/job/Hodgepodge/483/artifact/build/libs/hodgepodge-2.2.4-3-gd05ea26.jar

**Comment by Tommeh1** [comment]
> Huh maybe that utility class wasn't in the old gtnhmixin. try upgrade gtnhmixins to unimixin found in 2.3.2 zip, or just use this build instead: http://jenkins.usrv.eu:8080/job/Hodgepodge/483/artifact/build/libs/hodgepodge-2.2.4-3-gd05ea26.jar

Works now:
```java
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: com.mitchej123.hodgepodge.util.NBTTagCompoundConcurrentModificationException: Keys: Level...TileEntities...[id=BlockDrive]inv...item0...tag. Source tag: {#52:{Craft:0b,Cnt:8L,id:10573s,Count:0b,Damage:9s,Req:0L},#53:{Craft:0b,Cnt:610L,id:7191s,Count:0b,Damage:2034s,Req:0L},#54:{Craft:0b,Cnt:222L,id:7191s,Count:0b,Damage:2057s,Req:0L},#55:{Craft:0b,Cnt:80L,id:7191s,Count:0b,Damage:2835s,Req:0L},#56:{Craft:0b,Cnt:128L,id:7191s,Count:0b,Damage:2936s,Req:0L},#57:{Craft:0b,Cnt:64L,id:10567s,Count:0b,Damage:10s,Req:0L},#58:{Craft:0b,Cnt:26L,id:10567s,Count:0b,Damage:24s,Req:0L},#59:{Craft:0b,Cnt:21L,id:10573s,Count:0b,Damage:8s,Req:0L},upgrades:{},ic:8393L,#60:{Craft:0b,Cnt:174L,id:7191s,Count:0b,Damage:2830s,Req:0L},#61:{Craft:0b,Cnt:64L,id:2097s,Count:0b,Damage:839s,Req:0L},#62:{Craft:0b,Cnt:64L,id:3513s,Count:0b,Damage:28s,Req:0L},it:63s,@10:29L,@11:5L,@12:390L,@13:192L,@14:34L,@15:977L,@16:8L,@17:3L,@18:4L,@19:21L,#0:{Craft:0b,Cnt:10L,id:7191s,Count:0b,Damage:2804s,Req:0L},#1:{Craft:0b,Cnt:900L,id:10567s,Count:0b,Damage:28s,Req:0L},#2:{Craft:0b,Cnt:4L,id:7191s,Count:0b,Damage:2919s,Req:0L},#3:{Craft:0b,Cnt:23L,id:7191s,Count:0b,Damage:2848s,Req:0L},#4:{Craft:0b,Cnt:9L,id:7191s,Count:0b,Damage:2508s,Req:0L},#5:{Craft:0b,Cnt:215L,id:7191s,Count:0b,Damage:2838s,Req:0L},#6:{Craft:0b,Cnt:65L,id:7191s,Count:0b,Damage:2822s,Req:0L},#7:{Craft:0b,Cnt:10L,id:7191s,Count:0b,Damage:2821s,Req:0L},#8:{Craft:0b,Cnt:7L,id:10567s,Count:0b,Damage:7s,Req:0L},#9:{Craft:0b,Cnt:2L,id:7191s,Count:0b,Damage:2942s,Req:0L},config:{},@20:223L,@21:4L,@22:27L,@23:54L,#10:{Craft:0b,Cnt:29L,id:7191s,Count:0b,Damage:2842s,Req:0L},@24:212L,#11:{Craft:0b,Cnt:5L,id:10567s,Count:0b,Damage:29s,Req:0L},@25:4L,#12:{Craft:0b,Cnt:390L,id:7191s,Count:0b,Damage:2948s,Req:0L},@26:34L,#13:{Craft:0b,Cnt:192L,id:9650s,Count:0b,Damage:0s,Req:0L},@27:64L,#14:{Craft:0b,Cnt:34L,id:7191s,Count:0b,Damage:2871s,Req:0L},@28:4L,#15:{Craft:0b,Cnt:977L,id:7191s,Count:0b,Damage:2937s,Req:0L},@29:256L,#16:{Craft:0b,Cnt:8L,id:7191s,Count:0b,Damage:2914s,Req:0L},#17:{Craft:0b,Cnt:3L,id:10567s,Count:0b,Damage:26s,Req:0L},#18:{Craft:0b,Cnt:4L,id:7191s,Count:0b,Damage:2924s,Req:0L},#19:{Craft:0b,Cnt:21L,id:7191s,Count:0b,Damage:2524s,Req:0L},@30:211L,@31:128L,@32:128L,@33:64L,#20:{Craft:0b,Cnt:223L,id:7191s,Count:0b,Damage:2820s,Req:0L},@34:32L,#21:{Craft:0b,Cnt:4L,id:7191s,Count:0b,Damage:2904s,Req:0L},@35:32L,#22:{Craft:0b,Cnt:27L,id:7191s,Count:0b,Damage:2831s,Req:0L},@36:96L,#23:{Craft:0b,Cnt:54L,id:7191s,Count:0b,Damage:2917s,Req:0L},@37:32L,#24:{Craft:0b,Cnt:212L,id:7191s,Count:0b,Damage:2528s,Req:0L},@38:64L,#25:{Craft:0b,Cnt:4L,id:7191s,Count:0b,Damage:2873s,Req:0L},@39:48L,#26:{Craft:0b,Cnt:34L,id:7191s,Count:0b,Damage:2849s,Req:0L},#27:{Craft:0b,Cnt:64L,id:9652s,Count:0b,Damage:0s,Req:0L},#28:{Craft:0b,Cnt:4L,id:10567s,Count:0b,Damage:25s,Req:0L},#29:{Craft:0b,Cnt:256L,id:7191s,Count:0b,Damage:2855s,Req:0L},@40:32L,@41:32L,@42:32L,@43:64L,#30:{Craft:0b,Cnt:211L,id:3513s,Count:0b,Damage:23s,Req:0L},@44:64L,#31:{Craft:0b,Cnt:128L,id:2097s,Count:0b,Damage:59s,Req:0L},@45:128L,#32:{Craft:0b,Cnt:128L,id:3513s,Count:0b,Damage:5s,Req:0L},@46:1557L,#33:{Craft:0b,Cnt:64L,id:2097s,Count:0b,Damage:831s,Req:0L},@47:78L,#34:{Craft:0b,Cnt:32L,id:2097s,Count:0b,Damage:98s,Req:0L},@48:64L,#35:{Craft:0b,Cnt:32L,id:2097s,Count:0b,Damage:319s,Req:0L},@49:64L,#36:{Craft:0b,Cnt:96L,id:2097s,Count:0b,Damage:538s,Req:0L},#37:{Craft:0b,Cnt:32L,id:2097s,Count:0b,Damage:810s,Req:0L},#38:{Craft:0b,Cnt:64L,id:2097s,Count:0b,Damage:906s,Req:0L},#39:{Craft:0b,Cnt:48L,id:2097s,Count:0b,Damage:945s,Req:0L},@50:128L,@51:64L,@52:8L,@53:610L,#40:{Craft:0b,Cnt:32L,id:2097s,Count:0b,Damage:530s,Req:0L},@54:222L,#41:{Craft:0b,Cnt:32L,id:2097s,Count:0b,Damage:931s,Req:0L},@55:80L,#42:{Craft:0b,Cnt:32L,id:2097s,Count:0b,Damage:918s,Req:0L},@56:128L,#43:{Craft:0b,Cnt:64L,id:3513s,Count:0b,Damage:9s,Req:0L},@57:64L,#44:{Craft:0b,Cnt:64L,id:2097s,Count:0b,Damage:505s,Req:0L},@58:26L,#45:{Craft:0b,Cnt:128L,id:3513s,Count:0b,Damage:16s,Req:0L},@59:21L,#46:{Craft:0b,Cnt:1557L,id:7191s,Count:0b,Damage:2870s,Req:0L},#47:{Craft:0b,Cnt:78L,id:7191s,Count:0b,Damage:2414s,Req:0L},#48:{Craft:0b,Cnt:64L,id:2097s,Count:0b,Damage:855s,Req:0L},#49:{Craft:0b,Cnt:64L,id:2097s,Count:0b,Damage:936s,Req:0L},@0:10L,@1:900L,@2:4L,@3:23L,@4:9L,@5:215L,@6:65L,@7:10L,@8:7L,@9:2L,@60:174L,@61:64L,@62:64L,#50:{Craft:0b,Cnt:128L,id:7191s,Count:0b,Damage:2834s,Req:0L},#51:{Craft:0b,Cnt:64L,id:7191s,Count:0b,Damage:2923s,Req:0L}}
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.redirect$zco000$hodgepodge$checkCME(NBTTagCompound.java:538)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:29)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:376)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.redirect$zco000$hodgepodge$appendKeyPath(NBTTagCompound.java:554)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:31)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:376)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.redirect$zco000$hodgepodge$appendKeyPath(NBTTagCompound.java:554)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:31)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:376)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.redirect$zco000$hodgepodge$appendKeyPath(NBTTagCompound.java:554)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:31)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagList.func_74734_a(NBTTagList.java:32)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:376)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.redirect$zco000$hodgepodge$appendKeyPath(NBTTagCompound.java:554)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:31)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:376)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.redirect$zco000$hodgepodge$appendKeyPath(NBTTagCompound.java:554)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:31)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.CompressedStreamTools.func_150663_a(CompressedStreamTools.java:149)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.CompressedStreamTools.func_74800_a(CompressedStreamTools.java:139)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.world.chunk.storage.AnvilChunkLoader.func_75821_a(AnvilChunkLoader.java:265)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.world.chunk.storage.AnvilChunkLoader.func_75814_c(AnvilChunkLoader.java:251)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.world.storage.ThreadedFileIOBase.func_75736_b(SourceFile:30)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.world.storage.ThreadedFileIOBase.run(SourceFile:23)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at java.lang.Thread.run(Thread.java:748)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: Caused by: java.util.ConcurrentModificationException
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at java.util.HashMap$HashIterator.nextNode(HashMap.java:1445)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at java.util.HashMap$KeyIterator.next(HashMap.java:1469)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.redirect$zco000$hodgepodge$checkCME(NBTTagCompound.java:536)
[16:42:33] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	... 24 more
```

**Comment by Tommeh1** [comment]
Still exists in 2.3.2

**Comment by firenoo** [comment]
The affected tag actually looks like a fluid storage bus, but it's difficult to tell for sure with the information provided. Can the debug output print the full tag, not just a truncated version? @Glease

**Comment by Glease** [comment]
indeed. it's probably an item drive... I *can* make it dump the full tag, but the result will be near unreadable. Plus, the larger the tag to dump, the chance for the same concurrent modification error to happen during dumping would be greater too. (saving to disk is just a compact form of dumping and to a different target). 

in any case, here is one for you http://jenkins.usrv.eu:8080/job/Hodgepodge/485/artifact/build/libs/hodgepodge-2.2.4-5-gc08ef67.jar

**Comment by Tommeh1** [comment]
> The affected tag looks like some item drive type (not fluid), but it's difficult to tell for sure with the information provided. Can the debug output print the full tag, not just a truncated version? @Glease
> 
> Perhaps the new quantum cell.

We aren't using the quantum cell just yet, but i will give that hodgepodge build a go.

**Comment by chochem** [comment]
quantum isnt exactly new anymore. and only has one item type while the ones here have 63 anyway

**Comment by firenoo** [comment]
In that case its likely not ae2fc. I was thinking of one of the absurdly large cells that got a nerf from 63 -> 1. I don't think any item cell holds 63 types in the mod, AE2 (or Extra Cells) has them all at this point. AE2 has not had an update to drive serialization code since... 8 years ago?

**Comment by Tommeh1** [comment]
The debug logging returns null coordinates in the tag and the rest is over 65k characters long:
[CME tags.txt](https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/files/11183757/CME.tags.txt)

**Comment by Tommeh1** [comment]
The coordinates in the tag blame a drive containing only 1024k and 16384k drives, and the 16384k ones have only been added in yesterday
![image](https://user-images.githubusercontent.com/39203645/230734594-9db73c70-a17e-44fb-a9e0-5855145604b7.png)
[CME tags.txt](https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/files/11183763/CME.tags.txt)

**Comment by firenoo** [comment]
Some checklist items to test:
- Crafting system - is there a crafting job putting items in that ME (sub)system?
- Cells themselves - does removing the void cell fix things?
- Addons from other mods - do they try to insert directly into a drive and change the tag?

Finally, are we even sure that the tag being blamed is correct?

**Comment by Tommeh1** [comment]
> Some checklist items to test:
> 
> * Crafting system - is there a crafting job putting items in that ME (sub)system?
> * Cells themselves - does removing the void cell fix things?
> * Addons from other mods - do they try to insert directly into a drive and change the tag?
> 
> Finally, are we even sure that the tag being blamed is correct?

1 thing i can already say for sure and that is that neither of those systems have any CPUs, they're purely for processing raw materials.

**Comment by Tommeh1** [comment]
> Any GT ME hatches/buses? Which ones, configuration, etc?

All ME systems have ME stocking in and output busses, but not all cause CMEs.
And in terms of configuration i wouldn't say any systems are special in this regard, if anything this should end in the main system causing issues as well.

**Comment by firenoo** [comment]
The nature of how CMEs happen mean that they won't be consistent where they happen.

The file in question has a tag compound exposed: https://github.com/GTNewHorizons/Applied-Energistics-2-Unofficial/blob/master/src/main/java/appeng/me/storage/CellInventory.java#L44

Someone is modifying this tag (probably itself since its private, or someone is doing bad things with reflection...)

**Comment by Glease** [comment]
It now appears to me, that if 
1. your save is large enough OR your disk is slow enough
2. your ME system is super active

that CME can happen with AE2 alone. Consider the following chain of events
1. World save happen. Data is saved to a in-memory NBT structure
2. Write to disk happen. Data is now being iterated over.
2. Server tick happen. Someone put stuff into ME drive (or retrieve). AE2 eagerly save the change to ItemStack tag compound. This will modify the in-memory NBT structure. NOTE this happens in parallel to write to disk and its iteration over NBT.
3. Java found out the iteration is interrupted by that modification. CME is thrown.

TBH, this seems to be an inherent problem to the minecraft code and not just AE2 alone. It does not make a copy of item tag when writing to disk in general. AE2 is the top offender because it is the few thing that will modify the item nbt tag. If you setup a research station to repeatedly research stuff and the timing is correct (which is 100000000x less likely than AE2's case) it can cause CME too. Maybe some of those weird hard to debug lag-related dupe exploits are actually caused by this issue.

We now have several ways to solve this

1. retry write upon CME. This means a dirty write and the save might not be in a consistent state. In the above scenario, if one inspect the on disk save file, that extra stuff put into ME drive will be present in ME drive, but will not be removed from player inventory. 
2. Flag NBT tags and postpone writes to them until the flag is cleared, or make a copy when flag is present. This can be lag intensive and need great care to make it performant and correct.
3.  AE2 is the top offender, so fixing it should fix CME for the 99% of the case. Postpone writes to AE2 cells tag compound only until 
	* cell is removed from drive/chest
	* chunk save

**Comment by firenoo** [comment]
How feasible would it be to force MC to make a copy of the requested tag before saving the copy (if it isn't already), then tossing that copy when its written to disk? i.e.
```
for each NBTTagCompound tag:
    let x = tag.copy()
    x.writeTo(saveFile)
```

**Comment by Glease** [comment]
That makes save pause longer. Not sure how long that would be though.

**Comment by Tommeh1** [comment]
> TBH, this seems to be an inherent problem to the minecraft code and not just AE2 alone. It does not make a copy of item tag when writing to disk in general. AE2 is the top offender because it is the few thing that will modify the item nbt tag. If you setup a research station to repeatedly research stuff and the timing is correct (which is 100000000x less likely than AE2's case) it can cause CME too. Maybe some of those weird hard to debug lag-related dupe exploits are actually caused by this issue.

This would make sense given that like i've stated earlier that the 2 most affected systems are 2 with very high throughput.
In regards to the save however it is only roughly 5GB large and is stored on some beefy datacenter SSD.

Given your explanation i also wouldn't be surprised if that is what allowed #12891 to occur given that i fail to artificially trigger it again in any way.

**Comment by Glease** [comment]
sorry for going MIA for ~a week. just a status update. I don't currently have time to active work on this. IRL matters strike, and my attention on GTNH matter is mostly focused on the recent packet forging exploit patch up effort. This one is still on my backlog but that one will take priority.

**Comment by Tommeh1** [comment]
> sorry for going MIA for ~a week. just a status update. I don't currently have time to active work on this. IRL matters strike, and my attention on GTNH matter is mostly focused on the recent packet forging exploit patch up effort. This one is still on my backlog but that one will take priority.

Packet forging should have been looked into much earlier honestly, given that it was already known last year that validation of received data was potentially lacking as discovered in #12034

**Comment by boubou19** [comment]
> > sorry for going MIA for ~a week. just a status update. I don't currently have time to active work on this. IRL matters strike, and my attention on GTNH matter is mostly focused on the recent packet forging exploit patch up effort. This one is still on my backlog but that one will take priority.
> 
> Packet forging should have been looked into much earlier honestly, given that it was already known last year that validation of received data was potentially lacking as discovered in #12034

You don't even know what you are talking about and yet your allow yourself to criticize this. Really when will you get that we work on what we can/want, when we can/want. FYI it's much deeper than what you link in your message. I'm fed up with your attitude @Tommeh1 last warning.

**Comment by Tommeh1** [comment]
> > > sorry for going MIA for ~a week. just a status update. I don't currently have time to active work on this. IRL matters strike, and my attention on GTNH matter is mostly focused on the recent packet forging exploit patch up effort. This one is still on my backlog but that one will take priority.
> > 
> > 
> > Packet forging should have been looked into much earlier honestly, given that it was already known last year that validation of received data was potentially lacking as discovered in #12034
> 
> You don't even know what you are talking about and yet you allow yourself to criticize this. Really when will you get that we work on what we can/want, when we can/want. FYI it's much deeper than what you link in your message. I'm fed up with your attitude @Tommeh1 last warning.

I didn't mean anything bad with my statement.
It's simply that, to my understanding, if a malformed packet can crash a server, perhaps that should have hinted that manipulating it on purpose might also be possible as for as far as i know, both of these hint at a lack of validation of what data is received.

Neither do i say that someone must have been working on it a while ago, but it simply seems rather severe to me and thus working on it earlier would have been logical if the way i understand things as just stated is correct.

And if i'm wrong, feel free to correct me, but once more, i have no malicious intend with that message.

**Comment by Tommeh1** [comment]
Still exists in 2.3.4

**Comment by Loreili** [comment]
I'm think bumping in to the same behaviour on 2.3.4 although my stack trace is slightly different (few lines out, similar classes). I tried adding ` B:chunkSaveCMEDebug=true` to the hodgePodge config file in the debug section but I'm not seeing any additional logging - not sure what I've missed there.

`[08:55:04] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: java.util.ConcurrentModificationException
[08:55:04] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at java.util.HashMap$HashIterator.nextNode(HashMap.java:1469)
[08:55:04] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at java.util.HashMap$KeyIterator.next(HashMap.java:1493)
[08:55:04] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:29)
[08:55:04] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:376)
[08:55:04] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:31)
[08:55:04] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagList.func_74734_a(NBTTagList.java:32)
[08:55:04] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:376)
[08:55:04] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:31)
[08:55:04] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagList.func_74734_a(NBTTagList.java:32)
[08:55:04] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:376)
[08:55:04] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:31)
[08:55:04] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:376)
[08:55:04] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:31)
[08:55:04] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.CompressedStreamTools.func_150663_a(CompressedStreamTools.java:149)
[08:55:04] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.CompressedStreamTools.func_74800_a(CompressedStreamTools.java:139)
[08:55:04] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.world.chunk.storage.AnvilChunkLoader.func_75821_a(AnvilChunkLoader.java:265)
[08:55:04] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.world.chunk.storage.AnvilChunkLoader.func_75814_c(AnvilChunkLoader.java:251)
[08:55:04] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.world.storage.ThreadedFileIOBase.func_75736_b(SourceFile:30)
[08:55:04] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.world.storage.ThreadedFileIOBase.run(SourceFile:23)
[08:55:04] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at java.lang.Thread.run(Thread.java:750)`

What would be useful information to gather to help with any investigation?

I've now seen this in two separate chunks. Both have a bunch of Large Chemical Reactors in them but otherwise they're a significant distance apart (~1k blocks) so I would assume they're in different .rca world files.

**Comment by miozune** [comment]
@Loreili Use this build http://jenkins.usrv.eu:8080/job/Hodgepodge/541/ `chunkSaveCMEDebug` feature has been experimental and never merged to main.

**Comment by Loreili** [comment]
Thanks @miozune.

Since adding that patch and enabling the debug mode, I'm not seeing the same CME any more, just a stack of NPE errors. I saw the following on server `stop` after adding some multiblocks (LCRs) back in to a chunk which I've been seeing rollbacks on:

```
[07:36:51] [Server thread/DEBUG] [FML/]: Gathering id map for writing to world save World
[07:36:52] [Server thread/INFO] [STDOUT/]: [rwg.world.RwgWorldSavedData:func_76187_b:39]: RWG SAVING DATA DATA - NOISE: PERLIN
[07:36:52] [Server thread/INFO] [STDOUT/]: [rwg.world.RwgWorldSavedData:func_76187_b:40]: RWG SAVING DATA DATA - NOISE: PERLIN
[07:37:00] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: java.lang.NullPointerException
[07:37:00] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagIntArray.func_74734_a(NBTTagIntArray.java:22)
[07:37:00] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:376)
[07:37:00] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.redirect$zcm000$hodgepodge$appendKeyPath(NBTTagCompound.java:552)
[07:37:00] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:31)
[07:37:00] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:376)
[07:37:00] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.redirect$zcm000$hodgepodge$appendKeyPath(NBTTagCompound.java:552)
[07:37:00] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:31)
[07:37:00] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagList.func_74734_a(NBTTagList.java:32)
[07:37:00] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:376)
[07:37:00] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.redirect$zcm000$hodgepodge$appendKeyPath(NBTTagCompound.java:552)
[07:37:00] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:31)
[07:37:00] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:376)
[07:37:00] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.redirect$zcm000$hodgepodge$appendKeyPath(NBTTagCompound.java:552)
[07:37:00] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:31)
[07:37:00] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.CompressedStreamTools.func_150663_a(CompressedStreamTools.java:149)
[07:37:00] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.CompressedStreamTools.func_74800_a(CompressedStreamTools.java:139)
[07:37:00] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.world.chunk.storage.AnvilChunkLoader.func_75821_a(AnvilChunkLoader.java:265)
[07:37:00] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.world.chunk.storage.AnvilChunkLoader.func_75814_c(AnvilChunkLoader.java:251)
[07:37:00] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.world.storage.ThreadedFileIOBase.func_75736_b(SourceFile:30)
[07:37:00] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.world.storage.ThreadedFileIOBase.run(SourceFile:23)
[07:37:00] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at java.lang.Thread.run(Thread.java:750)
[07:37:37] [Server thread/DEBUG] [FML/]: Gathering id map for writing to world save World
[07:37:41] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: java.lang.NullPointerException
[07:37:41] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagIntArray.func_74734_a(NBTTagIntArray.java:22)
[07:37:41] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:376)
[07:37:41] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.redirect$zcm000$hodgepodge$appendKeyPath(NBTTagCompound.java:552)
[07:37:41] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:31)
[07:37:41] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:376)
[07:37:41] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.redirect$zcm000$hodgepodge$appendKeyPath(NBTTagCompound.java:552)
[07:37:41] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:31)
[07:37:41] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagList.func_74734_a(NBTTagList.java:32)
[07:37:41] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:376)
[07:37:41] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.redirect$zcm000$hodgepodge$appendKeyPath(NBTTagCompound.java:552)
[07:37:41] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:31)
[07:37:41] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:376)
[07:37:41] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.redirect$zcm000$hodgepodge$appendKeyPath(NBTTagCompound.java:552)
[07:37:41] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:31)
[07:37:41] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.CompressedStreamTools.func_150663_a(CompressedStreamTools.java:149)
[07:37:41] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.nbt.CompressedStreamTools.func_74800_a(CompressedStreamTools.java:139)
[07:37:41] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.world.chunk.storage.AnvilChunkLoader.func_75821_a(AnvilChunkLoader.java:265)
[07:37:41] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.world.chunk.storage.AnvilChunkLoader.func_75814_c(AnvilChunkLoader.java:251)
[07:37:41] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.world.storage.ThreadedFileIOBase.func_75736_b(SourceFile:30)
[07:37:41] [File IO Thread/INFO] [STDERR/]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.world.storage.ThreadedFile
```

There were two chunks where I was seeing this issue and after removing the GT multiblocks, wiring and AE2 from that chunk, it now saves and updates as I expected.

Shall I raise this as a separate issue and reference this one to avoid polluting these comments with possibly unrelated bumpf as the CME is not occurring here?

**Comment by miozune** [comment]
What version are you on? LCR NPE with `NBTTagIntArray` sounds like already fixed by https://github.com/GTNewHorizons/GT5-Unofficial/pull/2092 (GT 5.09.43.93)

**Comment by Loreili** [comment]
2.3.4 with v5.09.43.85 (pre-fix) as per https://gtnh.miraheze.org/wiki/Dev_Release_Stability

> rather badly incorrect fluid recipes. including broken DT and distilling recipes. highly recommended to update GT to [5.09.43.85](https://github.com/GTNewHorizons/GT5-Unofficial/releases/tag/5.09.43.85).

Any issues with bumping just that JAR up to .93 or would it be better to upgrade to something more recent (e.g. 2.3.6)?
